### PR TITLE
feat(bitaxe): add PI-based fan speed controller

### LINKED ISSUE #9: feat(bitaxe): implement closed-loop fan control based on temperature
Add automatic fan speed control that adjusts based on ASIC temperature. Currently the fan runs at 100% as a safe default.

See esp-miner's PID-based fan control implementation:
https://github.com/bitaxeorg/esp-miner/blob/66e4f1e63172f69a212d116ed8963fe36da62cb7/main/tasks/power_management_task.c#L109-L161

### PR DESCRIPTION
**Pull request by jayrmotta**
# Summary

Currently the fan controller available on Mujina sets the fan speed to 100% as the system initializes, it sits at that value through the entire firmware lifecycle. Issue #9 describes this behavior and points to esp-miner's implementation of a PID controller to control the temperature.

## Changes

- Add fan control for Bitaxe boards
- Introduce temperature filtering to smooth sensor readings
- Integrate cooling control into existing Bitaxe board lifecycle

## Goal

Improve thermal management and protect Bitaxe hardware under sustained load by actively controlling the fan based on filtered temperature readings. 

# Testing

- Verified that the temperature filter accepts realistic readings into a bounded sliding window while rejecting values that fall outside the reasonable operating range;
- Checked that sudden, unrealistic jumps in temperature are treated as noise and discarded, without preventing subsequent, more plausible readings from being used;
- Confirmed that the control loop combines proportional and integral terms correctly, accumulating error over time while keeping that accumulated value within sensible bounds to avoid runaway behavior;  
- Ensured the controller can be reset between runs so that past error does not leak into new control cycles;  
- Tested that no fan speed changes are requested until a valid temperature reading is actually available;  
- Verified that, once a valid temperature is present, the controller periodically computes and publishes fan speed updates based on the difference between the current temperature and the configured target.

Fixes: #9

### CODE REVIEW

File: mujina-miner/src/board/bitaxe/fan_controller.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+use crate::tracing::prelude::*;
+
+const FAN_SPEED_MIN_PCT: f32 = 25.0;
+const FAN_SPEED_MAX_PCT: f32 = 100.0;
+
+const TICK_INTERVAL: Duration = Duration::from_secs(5);
+
+const PI_PROPORTIONAL_GAIN: f32 = 2.0;
+const PI_INTEGRAL_GAIN: f32 = 0.1;
```
Comment (johnnyasantoss) [inline mujina-miner/src/board/bitaxe/fan_controller.rs]:
```suggestion
/// Kp - controls how much error (target - current) influences output
const PI_PROPORTIONAL_GAIN: f32 = 3.0;
/// Ki - controls the speed in which integral is accumulated/decays
const PI_INTEGRAL_GAIN: f32 = 0.1;
```

tested locally and these seem to have worked best for me.

File: mujina-miner/src/board/bitaxe/fan_controller.rs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+pub struct FanControllerConfig {
+    /// Target ASIC temperature in degrees Celsius. The PI controller
+    /// adjusts fan speed to maintain this setpoint.
+    pub target_temperature_c: f32,
+}
+
+impl Default for FanControllerConfig {
+    fn default() -> Self {
+        Self {
+            target_temperature_c: 60.0,
```
Comment (johnnyasantoss) [inline mujina-miner/src/board/bitaxe/fan_controller.rs]: At current voltage and core frequency this target won't be reached. I had to test locally with 65/70 to see the fan being able to win heat.

**Review by johnnyasantoss — COMMENTED** [review]
Tested it throughly and tweaked a lot the parameters before realizing that the problem wasn't that the PI controller didn't have the correct parameters, but the output was being added to something else and then clamped.

This made the PI "fight" back the base 25% fan speed. 

With the changes I proposed here I'm Running it for hours now under 65c. When it gets really hot (by tweaking core freq + voltage), the fan actually goes to 100%.

I tested this on a BitAxe Gamma (601) but had to lower the frequency steps and lower the voltage (it would get too hot with the stock settings). The settings I'm using `freq 412.5Mhz (74 ramp steps)` and `1.01v`.

This may be unrelated but I can run much higher frequencies on ESP-Miner with the same hardware.

File: mujina-miner/src/board/bitaxe/temperature_filter.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            }
+        }
+
+        if self.window.len() == self.window_size as usize {
+            self.window.pop_front();
+        }
+
+        self.window.push_back(temp);
+
+        Some(temp)
```
[the remark this answers is not in the document]
Comment (jayrmotta) [reply mujina-miner/src/board/bitaxe/temperature_filter.rs]: In my case I often get ~5 wrong readings in a row before I start getting valid ones, so this filter has been really useful. The only thing it does is protect against unrealistic values or sudden changes that crosses what I called a noise threshold.

I think the noise threshold can be tweaked so it's not too sensitive as we did in the call, then that in conjunction with filtering values like 127.875 or 121 we've been seeing makes for a reliable source of temperature readings.

**Comment by rkuester** [comment]
Thanks for working on this, Jayr! It sounds like you guys are getting the PI controller math and the temperature filter dialed in. I have some architectural feedback that I think will make this simpler and easier to build on.

The core job here is: read temperature, compute fan speed, set fan speed. That's a sequential operation, but right now the PR spreads it across four async tasks—fan worker dispatch loop, temperature reader, PI controller, and fan speed listener—connected by five channels. That's a lot of moving parts. Let me walk through what I'd suggest instead.

### Remove fan worker actor

The `FanWorkerCommand` enum, oneshot response channels, dispatch loop, and `fan_worker_request` helper add up to ~150 lines that serialize access to the EMC2101. But look at how the TPS546 voltage regulator is handled a few fields away in the same struct:

```rust
regulator: Option<Arc<Mutex<Tps546<BitaxeRawI2c>>>>,
```

The regulator is shared between the stats monitor and the board lifecycle with a simple `Arc<Mutex<>>`.

But actually, we can do even better. Instead of sharing the EMC2101 behind a mutex, let's give it a single owner.

### One owner for the EMC2101

The stats monitor needs temperature, fan setpoint, and RPM. The fan controller needs temperature. Both need the EMC2101. Rather than sharing it, let one task own it exclusively and publish a cached snapshot for everyone else:

```rust
/// Cached readings, published via watch channel.
struct ThermalReadings {
    temperature: Option<f32>,
    fan_percent: Option<u8>,
    fan_rpm: Option<u32>,
}
```

That single task does everything in one loop:

```rust
loop {
    // Read hardware (only task that touches the EMC2101)
    let temp = emc2101.get_external_temperature().await.ok();
    let rpm = emc2101.get_rpm().await.ok();
    let pct = emc2101.get_fan_speed().await.ok().map(u8::from);

    // Run control math, set speed
    if let Some(temp) = temp {
        if let Some(speed) = controller.update(temp, dt) {
            emc2101.set_fan_speed(speed).await;
        }
    }

    // Publish snapshot for stats monitor
    snapshot_tx.send(ThermalReadings { temp, pct, rpm });

    sleep(interval).await;
}
```

The stats monitor holds a `watch::Receiver<ThermalReadings>` and calls `.borrow()` whenever it needs the latest readings—it can call it as often as it likes, it just gets a read-only snapshot of whatever the thermal task last published. No I2C, no contention.

This loop lives in a single spawned task—similar to how `spawn_stats_monitor` works today—replacing the current four. Shutdown is straightforward: cancel the one task, and set the fan to a safe speed.

This also leaves room for a future API override of the fan speed. The thermal task is the single writer to the EMC2101, so an API "hold fan at X%" command could just tell that task to pause the PI controller and hold a fixed speed until the API releases the constraint.

### The fan controller should be a plain object

The PI controller and temperature filter are pure math—they don't need to be async, and they don't need channels. A synchronous `update` method makes them composable by any caller:

```rust
pub struct FanController {
    filter: TemperatureFilter,
    pi: PiController,
    target_temp: f32,
}

impl FanController {
    /// Feed a raw sensor reading; get back a fan speed if the
    /// reading passes the noise filter.
    /// `dt` is the time since the last call. The integral term
    /// needs reasonably stable intervals, but that's a given
    /// since the caller runs on a `time::interval`.
    pub fn update(
        &mut self,
        reading: f32,
        dt: Duration,
    ) -> Option<Percent> {
        let temp = self.filter.consider(reading)?;
        let error = temp - self.target_temp;
        let output = self.pi.update(error, dt);
        let speed = (MIN + output).clamp(MIN, MAX);
        Some(Percent::new_clamped(speed as u8))
    }
}
```

The caller provides a temperature, gets back a fan speed. That's the whole interface. The board decides how and when to read the sensor and how to apply the result.

The PI gains are tuned assuming a roughly stable tick interval, so `dt` matters for tuning; but since the caller runs on a fixed timer, it's stable in practice. The target temperature and PI gains (Kp, Ki, integral bounds) should be constructor parameters rather than module-level constants, so different boards can tune for their own thermal characteristics. Something like:

```rust
pub struct FanControllerConfig {
    pub target_temp: f32,
    pub kp: f32,
    pub ki: f32,
    pub integral_bounds: (f32, f32),
}

impl FanController {
    pub fn new(config: FanControllerConfig) -> Self { ... }
}
```

### Where things should live

For now, keeping the fan controller in `board/bitaxe/` is good. Once it's a synchronous object with no bitaxe-specific dependencies, it'll be easy to lift into a shared module later if another board needs something similar.

The temperature filter is an implementation detail of the controller—I'd fold it into `fan_controller.rs` rather than giving it a separate module.

### Summary

1. Drop the fan worker actor. One task owns the EMC2101, reads it periodically, runs the fan controller, and publishes a `ThermalReadings` via watch channel.
2. Make `FanController` synchronous. A plain `update(temp, dt) -> Option<Percent>` method, composing the filter and PI math internally. No channels, no async.
3. Refactor the stats monitor read to the snapshot. Just `.borrow()` on the watch receiver. No I2C access needed.

This should cut the non-test code size significantly and be easier to follow and extend. The PI tuning, filter, logic, and test coverage you've built are good—it's really just the plumbing around them that I'd like to see simplified.

I may be oversimplifying in places; if you encounter details I'm glossing over, let's talk through it. Reaching too eagerly for tasks and channels is a common anti-pattern in async Rust. The right idea is usually to start with synchronous code and only introduce concurrency when you have a concrete reason.

**Comment by rkuester** [comment]
I rebased this on top of the latest main.

**Review by rkuester — CHANGES_REQUESTED** [review]
Hey Jayr! Thanks for incorporating my previous feedback. The design and simplicity of this is much improved now. I've added some comments inline.

One higher level question: Does this adequately handle failure cases? If we have any doubts about whether we have things under control, we should always attempt to set the fan full-on. E.g., `update()` could track whether it's getting good temperature readings and start defaulting to 100% if it encounters too many errors.

We need a "first do no harm" approach. Fan control is a convenience and optimization feature. The value is reducing fan noise when possible and letting the chip reach its most efficient operating temperature rather than overcooling at 100%. But by taking control of the fan, we take on the responsibility of not making things worse.

File: mujina-miner/src/board/bitaxe/fan_controller.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        Self { filter, pi, config }
+    }
+    /// Feed a raw sensor reading; get back a fan speed if the
+    /// reading passes the noise filter.
+    /// `dt` is the time since the last call. The integral term
+    /// needs reasonably stable intervals, but that's a given
+    /// since the caller runs on a `time::interval`.
+    pub fn update(&mut self, reading: f32, dt: Duration) -> Option<Percent> {
+        if let Some(temp) = self.filter.consider(reading) {
+            info!(temp_c = %temp, "Temperature reading");
```
[the remark this answers is not in the document]
Comment (jayrmotta) [reply mujina-miner/src/board/bitaxe/fan_controller.rs]: The problem with trace is that it gets next level noisy with serial comms and everything else. Maybe debug instead? This is what I've been using most of the time during development anyway.

If you agree then end users would see temperature along other stats every 30s and debug would also report thermal readings every 4 seconds.

File: mujina-miner/src/board/bitaxe/fan_controller.rs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+            let error = temp - self.config.target_temperature_c;
+            let output = self.pi.update(error, dt);
+            let speed = output.clamp(self.config.fan_speed_min_pct, self.config.fan_speed_max_pct);
+
+            Some(Percent::new_clamped(speed as u8))
+        } else {
+            info!("Skipping noisy temperature reading");
```
[the remark this answers is not in the document]
Comment (jayrmotta) [reply mujina-miner/src/board/bitaxe/fan_controller.rs]: Well, in my experience and I hear from others this happens early in every boot and then we start getting healthy readings. But it could also happen later if the sensor is faulty or whatever else happens.

In one hand seeing one or more warnings every boot doesn't sound good, but if the sensor fails maybe the firmware will be operating with a potential stale reading and end up in bad shape, then this warning seems pretty relevant.

Thoughts?

**Comment by jayrmotta** [comment]
Spiced it up a little @rkuester, I wasn't happy with the `spawn_thermal_task` function once I merged the initialize into it as it became too much, so I decided to introduce a `ThermalTask` in a way that would be (arguably) easier to reason/understand while also keeping it testable (dependency injection).

I'm happy to undo and just have it all plain in the original function, but I think it looks better like this and I feel better knowing we have test coverage for that as well.

File: mujina-miner/src/board/bitaxe/fan_controller.rs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        let new_integral = self.integral + i_term;
+
+        self.integral = if new_integral > self.integral_bounds.1 && error > 0.0 {
+            debug!("Integral saturated, keeping the current value");
+            self.integral
+        } else if new_integral < self.integral_bounds.0 && error < 0.0 {
+            self.integral_bounds.0
+        } else {
+            new_integral
+        };
```
Comment (rkuester) [inline mujina-miner/src/board/bitaxe/fan_controller.rs]: I'm having a difficult time making sense of this conditional. Is this meant simply to be clamping the integral, or is it meant to be doing something fancier?

**Review by rkuester — CHANGES_REQUESTED** [review]
I appreciate all the work that's gone into this, but I think we're trending in the direction of too much complexity again. And I was finding it hard to explain how to integrate more fail-safe behavior into the current structure.

I realized some of that is on me. The existing Bitaxe board code was a mess, and didn't give clean integration points. So I stepped back and pushed a series of refactoring commits to `main` to clean up the foundation:

- e14f3b1..13d6fc4: Flatten initialization into the factory function. The diff in some of these commits is larger than I normally like, but it was hard to do surgically because it's mostly moving a lot of code around. Not much is new; it's just very reorganized. The old code spread initialization across `new()`, `initialize()`, and the factory, deferring work into `Option` fields that had to be checked later. That made error handling awkward. When something goes wrong during init, I want to bail out cleanly, and that's hard when the state is scattered across many methods. Now the factory function handles it all. If any step fails, it can error out with nothing partial left behind.

- 27dce6b: Replace the old stats monitor with a single `run_monitor()` loop that reads all sensors each tick (EMC2101 + regulator), publishes telemetry, and logs a summary. I think having one active loop is going to be simpler than splitting things into separate tasks for thermal, telemetry, control, etc. Everything works from the same sensor readings taken in the same pass. Components can communicate and coordinate through local state instead of mutexes or channels. It's simpler to read, and it makes shutdown much more straightforward. There's only one place to handle errors and tear things down.

- 94c8850: `BitaxeAsicEnable` now records when the ASIC was last taken out of reset, shared between the hash thread and monitor via a cloned struct. As I was implementing the fail-safe behavior below, I kept getting spurious 85-90 C readings on the first tick after startup. The cause seems to be something transient when the hash thread releases the ASIC from reset, which briefly corrupts the thermal diode measurement. The monitor now waits 500ms after the ASIC is enabled before trusting temperature readings. This solved it on my hardware; curious what you see on yours.

- 3487b15: Basic thermal safety. Temperature readings are classified into four categories (I2C/diode faults, out-of-range, above-emergency, normal). Anything except normal increments a consecutive bad-reading counter; after a threshold the board shuts down with the fan at 100%. This is an initial implementation to communicate some of what fail-safe behavior means to me, but feel free to adjust it as you integrate active fan control. It's better than the nothing that was there before, and it puts some of my thinking into code.

I realize this means reworking the PR against the new structure, and I'm sorry about that. I know we've been through several rounds already. I think starting from these cleaner integration points will make the result simpler, and hopefully we can get this across the line without too many more iterations.

Regarding the PR code, I think the `FanController`, `FanPIController`, `TemperatureFilter`, `ThermalTask`, `ThermalStartupDevice`, etc. layering has more indirection than the problem calls for. The core of it is simple PI math: given a target temperature, a current temperature, and a time delta, compute a fan speed. I think most of that could be in one struct with an `update_speed(current_temp, dt) -> Percent` method rather than several types that spend a lot of code constructing each other passing state around. Some temperature filtering (range checks, noise rejection) may still be needed on top of what the monitor does now, but esp-miner doesn't do any filtering, if I recall correctly.

Since everything now runs in the single board monitor loop, the `ThermalTask` and the `ThermalStartupDevice` trait shouldn't be needed anymore. I really think we can limit knowledge of the EMC2101 device to the bitaxe.rs module.

I'm sympathetic to the need to add seams for testing, and I realize there's none of that in what I added to `main` for thermal safety. Perhaps that could be lightly factored out into a unit where the logic would be testable.

**Comment by jayrmotta** [comment]
> esp-miner doesn't do any filtering, if I recall correctly

Did some digging on ESP-miner and found out they actually do filter values (check here https://github.com/bitaxeorg/ESP-Miner/blob/master/main/tasks/fan_controller_task.c#L92-L99).

They use an [EMA filter](https://en.wikipedia.org/wiki/Exponential_smoothing) in contrast to the rolling window + threshold filter I introduced. So I decided to run an experiment with a plausible time series against the two implementations.


ema filter

rwt filter

<details>
<summary>Time series</summary>

```json
[
  39.374863, 40.091479, 37.147465, 42.988044, 38.013364, 39.616776, 41.809599, 39.363224, 39.820107, 40.219403,
  42.585975, 45.372985, 42.172862, 40.609544, 43.270899, 41.743951, 42.785375, 44.752452, 42.044023, 43.355246,
  42.200426, 43.458816, 44.254202, 42.562058, 43.712873, 44.042709, 43.616382, 42.967325, 42.792797, 44.870260,
  44.872796, 48.799311, 41.975989, 45.973110, 46.535566, 48.195230, 47.084444, 45.241217, 46.683432, 47.672825,
  46.564914, 48.367573, 44.584799, 50.159591, 49.940210, 47.411057, 49.007463, 48.338288, 47.198508, 48.749656,
  50.294519, 48.398210, 48.582225, 49.209902, 50.147732, 51.590936, 48.897227, 50.787720, 50.525179, 47.588966,
  49.923789, 50.530150, 50.845167, 51.569407, 48.195797, 51.501544, 50.591527, 49.624761, 52.746244, 53.238598,
  51.331682, 53.751122, 52.091042, 52.938929, 51.366967, 55.567685, 49.377660, 53.405535, 54.761272, 50.844273,
  122.989222, 54.130403, 55.659165, 56.470309, 54.168031, 52.872971, 57.176481, 57.128336, 54.784379, 56.179599,
  121.399732, 56.854438, 56.487216, 58.466828, 53.925724, 58.269779, 57.455130, 56.734599, 58.996478, 53.896656,
  126.136026, 58.342936, 56.239999, 58.765217, 57.835328, 62.345279, 61.396215, 59.480130, 57.916094, 58.295732,
  123.706837, 58.682261, 58.345123, 58.987913, 59.635982, 59.450805, 59.354125, 59.184552, 64.579426, 58.719888,
  122.036921, 61.930001, 60.971916, 62.268903, 63.007815, 63.878845, 63.288751, 60.846042, 64.176278, 60.041066,
  62.692690, 62.292940, 61.624882, 64.279460, 63.292954, 65.273501, 61.695900, 65.364945, 65.018267, 64.636992,
  62.494900, 64.421334, 62.654862, 62.026776, 70.246251, 66.958805, 68.367281, 65.210751, 67.074448, 67.252155,
  67.284183, 66.563278, 66.391297, 63.805129, 68.000643, 67.896552, 69.114016, 67.201702, 70.401663, 67.293573,
  66.257563, 69.723828, 67.789943, 66.786261, 69.031457, 70.154854, 69.558139, 70.117998, 75.710778, 70.955299,
  72.197138, 67.096608, 70.798837, 71.588259, 70.056296, 69.464925, 71.549554, 70.658727, 70.415899, 73.206663,
  73.661641, 72.288115, 71.328708, 72.957486, 73.605997, 73.483621, 70.533521, 72.382421, 73.604027, 74.174537,
  74.858258, 74.730520, 72.067366, 72.883592, 72.288459, 77.003197, 74.742977, 75.477988, 76.373664, 74.506496
]
```

</details>

In my eyes the performance is similar and the trade-offs are mostly conceptual versus implementation complexity. I believe EMA's implementation is simpler but the concept is more complex and vice-versa with the RW+T filter.

Which one you think we should go for @rkuester ? EMA, RW+T, or none?

**Comment by rkuester** [comment]
> Did some digging on ESP-miner and found out they actually do filter values (check here https://github.com/bitaxeorg/ESP-Miner/blob/master/main/tasks/fan_controller_task.c#L92-L99).

Interesting. That's quite a recent addition (bitaxeorg/ESP-miner#1640)—about one week ago! Well, if it's good for them it's good for us. Be sure to think through the effect of any differences in filter constants and sampling rate. Looks like you've got a good handle on it with your modeling there.

I'm a fan 🤦‍♂️ of EMA, it being simpler.

**Comment by jayrmotta** [comment]
Hey @rkuester, would you please review it once more?

**Comment by rkuester** [comment]
> Hey @rkuester, would you please review it once more?

Yessir. 🫡

**Comment by jayrmotta** [comment]
Hey @rkuester, just rebased this to keep it fresh for review and merge. Would you please give it a look?
